src/deny-imports.ts

```diff
@@ -0,0 +1,105 @@
+import path from "node:path"
+
+import { normalizePath, type PluginOption, type ResolvedConfig } from "vite"
+
+import pkg from "../package.json"
+import { type Env } from "./env"
+import { type Pattern, findMatch } from "./pattern"
+
+type EnvPatterns = Partial<Record<Env, Pattern[]>>
+
+type Options = Partial<
+  Record<Env, { specifiers?: Pattern[]; files?: Pattern[] }>
+>
+
+export default function (options: Options): PluginOption[] {
+  let denySpecifiers: EnvPatterns = {}
+  let denyFiles: EnvPatterns = {}
+
+  for (let [env, { specifiers, files }] of Object.entries(options)) {
+    denySpecifiers[env as Env] = specifiers
+    denyFiles[env as Env] = files
+  }
+
+  return [denyImportSpecifiers(denySpecifiers), denyImportFiles(denyFiles)]
```

Comment:
Splitting it out into two plugins is really clever, that keeps the code so much simpler.